core/src/main/java/io/grpc/transport/AbstractClientStream.java

```diff
@@ -45,9 +47,12 @@
 import java.util.logging.Level;
 import java.util.logging.Logger;
 
+import javax.annotation.concurrent.NotThreadSafe;
+
 /**
  * The abstract base class for {@link ClientStream} implementations.
  */
+@NotThreadSafe
```

Comment:
The @ ThreadSafe follows the practice that I described: "something isn't thread-safe unless it says it is thread-safe." It sounds like better commenting of AbstractStream is in order (maybe by using @ NotThreadSafe) instead of adding the annotation here.